src/qutip_cupy/dense.py

```diff
@@ -0,0 +1,130 @@
+"""" This module contains the ```CuPyDense``` class as well as all its 
+asociated function conversion and specializations functions as 
+required by QuTiP's Data Layer """
+
+import numbers
+
+import cupy as cp
+from qutip.core import data
+
+
+class CuPyDense(data.Data):
+    def __init__(self, data, shape=None, copy=True, dtype=cp.complex128):
+        """ 
+            This class provides a dense matrix backend for QuTiP.
+            Matrices are stored internally in a CuPy array on a GPU.
+            If you have many GPUs you can set GPU ``i`` 
+            by calling ``cp.cuda.Device(i).use()`` before construction.
+            
+            Parameters
+            ----------
+            data: array-like 
+                Data to be stored.
+            shape: (int, int) 
+                Defaults to ``None``. If ``None`` will infer the shape from ``data``,
+                else it will set the shape for the internal CuPy array. 
+            copy: bool 
+                Defaults to ``True``. Whether to make a copy of 
+                the elements in ``data`` or not.
+            dtype:
+                Data type specifier. Either ``cp.complex128`` or ``cp.complex64``
+            
+        """
+        self.dtype = dtype
+
+        base = cp.array(data, dtype=self.dtype, order='K', copy=copy)
+        if shape is None:
+            shape = base.shape
+            # Promote to a ket by default if passed 1D data.
+            if len(shape) == 1:
+                shape = (shape[0], 1)
+        if not (
+            len(shape) == 2
+            and isinstance(shape[0], numbers.Integral)
+            and isinstance(shape[1], numbers.Integral)
+            and shape[0] > 0
+            and shape[1] > 0
+        ):
+            raise ValueError("shape must be a 2-tuple of positive ints, but is " + repr(shape))
+        if shape and (shape[0] != base.shape[0] or shape[1] != base.shape[1]):
+            if shape[0] * shape[1] != base.size:
+                raise ValueError("".join([
+                    "invalid shape ",
+                    str(shape),
+                    " for input data with size ",
+                    str(base.size)
+                ]))
+            else:
+                self._cp = base.reshape(shape)
+        else:
+
+            self._cp = base
+
+        super().__init__((shape[0], shape[1]))
+
+    @classmethod
+    def _no_checks_constructor(cls, data):
+        # This constructor should only be called when we are sure the argument
+        # is a CuPy 2-dimensional array.
```

Comment:
```suggestion
    def _raw_cupy_constructor(cls, data):
        """
        A fast low-level constructor for wrapping an existing CuPy array in a CuPyDense object 
        without copying it.   
        
        The ``data`` argument must be a CuPy array with the correct shape.
        The CuPy array will not be copied and will be used as is. 
        """
```
I have a personal preference for putting important information into docstrings rather than comments when it's about a function as a whole.

I have a hunch that we'll want to expose this alternative constructor publicly at some point, but we'll need a better name for it and let's not do it now (we can do it later if it still seems like a good idea). The hunch is that it's very useful for us, so it will likely be very useful to someone else who has a CuPy array and knows what they're doing.

Reply:
b23aec8